src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -7050,22 +7086,48 @@ int RGWRados::Object::prepare_atomic_modification(const DoutPrefixProvider *dpp,
   }
 
   if (need_guard) {
-    /* first verify that the object wasn't replaced under */
-    if (if_nomatch == NULL || strcmp(if_nomatch, "*") != 0) {
-      op.cmpxattr(RGW_ATTR_ID_TAG, LIBRADOS_CMPXATTR_OP_EQ, state->obj_tag); 
+    RGWObjState* current_state = state;
+
+    using namespace std::string_literals;
+    if (handle_version) {
+      // objects in versioning-enabled buckets don't get overwritten.
+      // preconditions refer to the current version instead
+      rgw_obj current_obj = obj;
+      current_obj.key.instance.clear();
+      constexpr bool follow_olh = true; // look up current version
+      r = store->get_obj_state(dpp, &ctx, bucket_info, current_obj,
+                               &current_state, nullptr, follow_olh, y);
+      if (r == -ENOENT) {
+        ldpp_dout(dpp, 4) << "no current object version for preconditions on "
+            << current_obj << dendl;
+      } else if (r < 0) {
+        ldpp_dout(dpp, 4) << "failed to load current version for " << current_obj << dendl;
+        return r;
+      } else {
+        ldpp_dout(dpp, 4) << "loaded current object " << current_state->obj
+            << " for preconditions" << dendl;
+      }
+    } else if (if_nomatch == NULL || if_nomatch != "*"sv) {
+      /* first verify that the object wasn't replaced under */
+      op.cmpxattr(RGW_ATTR_ID_TAG, LIBRADOS_CMPXATTR_OP_EQ, state->obj_tag);
       // FIXME: need to add FAIL_NOTEXIST_OK for racing deletion
     }
 
     if (if_match) {
       if (strcmp(if_match, "*") == 0) {
         // test the object is existing
-        if (!state->exists) {
+        if (!current_state->exists) {
           return -ERR_PRECONDITION_FAILED;
```

Comment:
for delete markers, get_obj_state() will return ENOENT and `current_state` would still be pointing to the wrong object. in my example, the ENOENT case set `current_state = nullptr` so you could handle that case

p.s. this not-exist case is supposed to return ENOENT/404